src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -1818,6 +1825,145 @@ impl Coordinator {
         };
     }
 
+    async fn explain_pushdown_select(
+        &mut self,
+        ctx: &mut ExecuteContext,
+        plan: SelectPlan,
+        target_cluster: TargetCluster,
+    ) -> Result<Vec<Row>, AdapterError> {
+        let mut results = vec![];
+        let catalog = self.owned_catalog();
+        let target_cluster_id = catalog
+            .resolve_target_cluster(target_cluster, ctx.session())?
+            .id;
+        let compute_instance = self
+            .instance_snapshot(target_cluster_id)
+            .expect("compute instance does not exist");
+
+        // Pick a timestamp and construct a stats oracle
+        let ids = plan.source.depends_on();
+        let timeline_context = self.validate_timeline_context(ids.iter().cloned())?;
+
+        let oracle_ts = self
+            .oracle_read_ts(ctx.session(), &timeline_context, &plan.when)
+            .await;
+
+        let id_bundle = self
+            .dataflow_builder(compute_instance.instance_id())
+            .sufficient_collections(ids.iter());
+
+        let timestamp_determination = self
+            .determine_timestamp(
+                ctx.session(),
+                &id_bundle,
+                &plan.when,
+                compute_instance.instance_id(),
+                &timeline_context,
+                oracle_ts,
+                None,
+            )
+            .await?;
+
+        let as_of = timestamp_determination.timestamp_context.antichain();
+
+        // Construct a statistics oracle
+        let oracle = self
+            .statistics_oracle(ctx.session(), &ids, &as_of, true)
+            .await?;
+
+        // Collect optimizer parameters.
+        let select_id = self.allocate_transient_id()?;
+        let index_id = self.allocate_transient_id()?;
+        let system_config = catalog.system_config();
+        let optimizer_config = OptimizerConfig::from(system_config);
+        let mut optimizer = optimize::peek::Optimizer::new(
+            Arc::clone(&catalog),
+            compute_instance,
+            plan.finishing,
+            select_id,
+            index_id,
+            optimizer_config.clone(),
+        );
+        let local_plan = optimizer.optimize(plan.source)?;
+        let resolved = local_plan.resolve(ctx.session(), oracle);
+        let global_plan = optimizer.optimize(resolved)?;
+
+        // We have the sources (including MFPs) and the corresponding timestamps; collect stats.
+        for (gid, (desc, _)) in global_plan.df_desc().source_imports.iter() {
```

Comment:
These imports seem to include non-Persist-sources. Is there a good way to filter this down, or does this get sorted out at a later optimization stage?

Reply:
I notice that if I optimize all the way down to LIR, I'm getting a slow-path plan in cases that should fall through to the fast path. I must be running the optimizer in some different way...